### PR TITLE
core-api: refactor and slim down the FeatureFlagsApi

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

The current feature flags API had a lot of concrete types mixed in, as well as a couple of layers making it awkward to use and implement. This distills the API down into 4 methods that covers all the existing use-cases, while leaving room for further evolving it in the future. For example adding another state is now possible, and it's still straight-forward to add observability.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/cost-insights/src/components/CostInsightsPage/CostInsightsPage.tsx [hunk 1]
```diff
@@ -211,7 +209,7 @@ export const CostInsightsPage = () => {
         </Typography>
       </Box>
       <Box minHeight={40} maxHeight={60} display="flex">
-        {!!flags.get('cost-insights-currencies') && (
```
Comment (Rugvip) [inline plugins/cost-insights/src/components/CostInsightsPage/CostInsightsPage.tsx:214]: Existing usage like this is not forwards compatible with all possible state additions, which is why there's only an `isActive()` now

**Review by timbonicus — APPROVED** [review]

**Comment by freben** [comment]
So nothing is async essentially. I understand that save can be done in the background, but do we need a blocking point at which the initial flags are loaded?

**Comment by benjdlambert** [comment]
@freben the`localStorage` API is synchronous, so everything is blocking in this PR. If we decide to move the storage to something like a remote service then I guess we would need that API. I guess it just depends on what the chance of that is

**Comment by freben** [comment]
yeah if we think we'll never want these to be stored remotely then it's ok of course

**Comment by Rugvip** [comment]
Yeah kept async flags out of it. The API is somewhat forwards-compatible with observing flags, but in general this is for manual feature switches.

**Review by freben — APPROVED** [review]

**Comment by Rugvip** [comment]
We may want to keep iterating on the API, but for now I think it's best to merge this and spend time elsewhere 😅

**Pull request merged**